Andrea Baretta  
Math 151AH  
Prof. Zhou  
1/28/2024  
Homework 1

1. The exact value of $1 − cos(x)$ when $x = 10^{−3}$ is
$$4.9999995833333472222219742063519620811078710385413014376488... × 10ˆ{-7}$$

1a. Using double precision in your programming language of choice, evaluate $1 − \cos \left(10^{−3}\right)$ and determine the number of significant figures that are accurate.

In [12]:
import numpy as np
double_precision = 1 - np.cos(1e-3)
print(double_precision)

4.999999583255033e-07


Compared to $$4.9999995833333472222219742063519620811078710385413014376488... × 10ˆ{-7}$$ it is accurate only to 4.999999583e-7, or 10 decimal places.

1b. Find a suitable trigonometric identity so that 1 − cos(x) can be accurately computed for |x| ≪ 1.
Your answer should ensure that there is no subtraction of numbers that are close to each other.

$$1-\cos(x)=1-\cos(x) \cdot \frac{1+\cos(x)}{1+\cos(x)}= \frac{\sin^2(x)}{1+\cos(x)}$$

1c. Use your answer from (b) to compute $1 − \cos(10^{−3})$ in double precision. How does it compare
with the result from (a)?

In [13]:
better_precision = (np.sin(1e-3)**2)/(1+np.cos(1e-3))
print(better_precision)

4.999999583333347e-07


Changing the formula to an equivalent form yielded substantially better results. Now, the final result is accurate to 16 decimal places, 6 more than before

2. Let $fl(x)$ denote the floating point approximation of a number, and assume that machine precision $\epsilon = 10^{−β}$ for some positive integer $\beta$.

2a. Consider evaluating the quantity $\sin\left(\frac{\pi}{2} + \delta\right)$ when $0 < \delta ≪ 1$. Use a Taylor expansion to give an estimate for the largest $\delta$ such that  
fl$\left(\sin\left(\frac{\pi}{2}\right)\right) =$ fl$\left(\sin\left(\frac{\pi}{2} + \delta\right)\right)$. (your answer should depend on $\beta$).

Since fl$\left(\sin\left(\frac{\pi}{2} + \delta\right)\right) =$ fl$\left(\cos\left(\delta\right)\right)$, consider the Taylor expansion of the cosine.
$$\cos(\delta) = 1-\frac{\delta^2}{2!}+\frac{\delta^4}{4!}+...$$
We want to ensure that all of the terms following the 1 end up evaluating to 0 in the floating point representation. Since the computer is accurate only to $\beta$ decimal places, we could take $\delta=10^{\beta/2}$, since then $\delta^2=10^{\beta}=\epsilon$. However, since that first term is divided by 2, we can double the size of $\delta$. Then, a reasonable estimate of the largest $\delta$ such that fl$\left(\sin\left(\frac{\pi}{2}\right)\right) =$ fl$\left(\sin\left(\frac{\pi}{2} + \delta\right)\right)$ is $\delta = 2\cdot10^{-\beta/2}$.

2c. Using double precision and your programming language of choice, determine the greatest $\delta$ for
which fl$\left(\sin\left(\frac{\pi}{2}\right)\right) =$ fl$\left(\sin\left(\frac{\pi}{2} + \delta\right)\right)$. (Your answer should be correct to 3 significant figures.)

In [18]:
upper_bound = 1.0
lower_bound = 0.0
n = 1000
for i in range(0, n):
    mid = (upper_bound + lower_bound)/2.0
    if np.sin(np.pi/2 + mid) == 1:
        lower_bound = mid
    else:
        upper_bound = mid
print(lower_bound)
print(np.sin(np.pi/2 + lower_bound))

1.0536712280462266e-08
1.0


Experimentally, increasing n beyond this point does not yield any different results. It is reasonable to believe, then, this algorithm has converged to the best possible value for $\delta$, $1.0536712280462266\cdot10^{-9}$.